### ISSUE TITLE
Remove boost spirit dependency

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2646]
Hello everyone, 

mlpack uses Boost libraries as a dependency that provides us with many features. Until recently, we have discovered that these dependencies are making mlpack harder to install, to maintain, and increasing mlpack footprint drastically as you can see here #2440 
Following the #2440 discussion, we have agreed to replace all Boost libraries one by one. Therefore we have started by replacing Boost program options with CLI11 in #2459, also we are working actively to replace the Boost serialization library by Cereal in #2458  and the Boost unit test framework in #2644 #2645 #2584 #2594, etc...

mlpack uses Boost spirit as a its main CSV parser, it will be a good idea to continue the above work and remove this one too.
Boost spirit provides mlpack with a fast parser. Therefore a replacement needs to be at least as fast as Boost spirit or probably better.

**Possible solution**:
- Replace boost spirit with a header-only library that can be included in one file
- Write from scratch a replacement for the functionality provided by boost sprit (CSV parser, etc)

If you have any possible solution, question, or interest in working on this issue please feel free to drop a comment on this issue so we can continue the discussion.

**Comment by Aakash-kaushik** [comment]
Hey @shrit I don't exactly know about any replacement libraries but if this discussion ends on everyone agreeing to reimplement functionalities similar to that are provided by boost spirit. I can help with that.

**Comment by rcurtin** [comment]
If possible I'd prefer to not implement an in-house CSV parser, but if we have to we have to.  Our existing boost::spirit parser is in `load_csv.cpp`---so really anything that matches the API provided by `LoadCSV()` should work okay.  Here are some options of out-of-the-box tools we could use:

https://github.com/ben-strasser/fast-cpp-csv-parser

https://github.com/d99kris/rapidcsv

Those are just two I found in a quick search.  Both are BSD 3-clause.  The disadvantage is that to release them with mlpack, we have a distribution problem:

 - If we include the header file directly in mlpack's source code, then it is hard to stay up to date with what happens upstream.

 - If we use an "autodownloader" like we do for ensmallen, we risk breakages if we run against the master branch of whatever we are wrapping.  Similarly, this makes the configuration process more complex, since we need to check locally for the dependency or download it if we can't find it.

 - If we don't include the header file in mlpack's source code and use an autodownloader, this can be problematic for some packaging situations.  A lot of distributions that package mlpack will require that the build is fully reproducible and doesn't depend on outside sources.  For instance, nix by default disables network access when building a package.
    - I'm the Fedora maintainer for mlpack, and what I would probably do is create an additional package for the CSV header.  Other distributions might do the same.  It's a bit tedious, so that's a decent bit of work for the maintainer.

So, there really is no silver bullet; if we handwrite a CSV parser, we have to pay the implementation time and maintenance cost, although we avoid the aforementioned distribution issues.

Anyway, it's all open for discussion---I'm not sure which solution is really better or worse; they both have advantages and disadvantages.  However, no matter what, one thing that will be really gratifying for the person working on this is timing the build *before* removing `boost::spirit` and then timing it *after* removing `boost::spirit`.  That should be a really nice benchmark to see. :smile:

**Comment by NippunSharma** [comment]
Hi, I would like to have a take on this. As @rcurtin mentioned the advantages and disadvantages of both sides, I would like to try implementing one and see if it performs same as boost::spirit (I doubt that given my little experience but I will perform some benchmarks). If there is any other information or resources you can provide that can be helpful ?

**Comment by NippunSharma** [comment]
Hi @rcurtin , @shrit I have started working on this. I have gone through the current codebase for reading csv (policies, dataset_mapper, load_csv). I have few a few questions: 

1. Should the new csv parser implementation 
    parse using PEG as boost spirit does? 

2. Should the new implementation have a same 
     structure/API as boost spirit? 

3. I noticed that armadillo also has a load method 
    that loads a csv file into a matrix. Does that 
    also use boost spirit for parsing? Can we use 
    armadillo in place of boost spirit?

**Comment by shrit** [comment]
> Hi @rcurtin , @shrit I have started working on this. I have gone through the current codebase for reading csv (policies, dataset_mapper, load_csv). I have few a few questions:
> 
>     1. Should the new csv parser implementation
>        parse using PEG as boost spirit does?
I am not sure about this one, maybe @rcurtin has a better answer here.

>     2. Should the new implementation have a same
>        structure/API as boost spirit?
It will be nice to keep the same API compatibility, especially for future users.

>     3. I noticed that armadillo also has a load method
>        that loads a csv file into a matrix. Does that
>        also use boost spirit for parsing? Can we use
>        armadillo in place of boost spirit?

In this case, it would be good to test the performance of the armadillo csv parser and compare it to our boost spirit parser. The objective is to implement a parser which have at least the same performance we have today. Armadillo does not require Boost, and the parser is implemented directly in it.

**Comment by NippunSharma** [comment]
@shrit thanks for answering. I will perform and share the benchmarks on the armadillo vs our boost spirit parser.

**Comment by NippunSharma** [comment]
Hi @shrit @rcurtin . I have performed benchmarks between mlpack's csv parser and armadillo's csv parser. I made csv files of different sizes. Each csv file was made from a square matrix with number of rows equal to `N`. `N` was varied from 1 to 5000.
According to the benchmark, both the parsers **perform very similar** to each other with **no one being a clear winner.**
Here is a plot I made from the benchmark results (the x-axis is `N` or the size of matrix from which csv file was made):
![image](https://user-images.githubusercontent.com/53967069/100008484-88c75600-2df3-11eb-88ef-e9e8645967b8.png)

**Comment by shrit** [comment]
@NippunSharma Thanks for sharing these results, great work.
I would like also to increase the size of `N.N` from `5000` to `1000000`, and another one from `10*6` to `10*8` the reason for this is that most of the machine learning datasets are larger than `5000` so we can not be sure only be testing only a small amount of dataset.
Would it be possible to do these tests?

**Comment by rcurtin** [comment]
@NippunSharma thanks for running these benchmarks!  In most cases, like @shrit pointed out, the number of rows in the CSV will be many, and at the same time the dimensionality (number of columns) should be reasonably low.

Can you show the code that you used?  If you call `data::Load(filename, matrix)`, it will use Armadillo's CSV parser internally (take a look at `src/mlpack/core/data/load_impl.hpp`).

To use the `boost::spirit` parser that we have implemented, you'll want to use the overload of `data::Load()` that takes a `DatasetInfo` or `DatasetMapper<PolicyType>`.  Specifically the reason for those classes is so that we can load a CSV (or other type) with non-numeric types in it.  The Armadillo CSV reader will fail to load a CSV like this:

```
1, 2, "hello", 4
2, 4, "goodbye", 5
```

So, we had to implement our own CSV parser to be able to parse strings (and convert them to, e.g., integer indexes or similar).  You can see how this works in `src/mlpack/tests/load_save_test.cpp`; for instance, in `LoadQuotedStringInCSVTest` there is a usage of this overload.  For a comparison of the boost::spirit parser to Armadillo's, you'd probably want to use a `DatasetInfo` on a CSV that contains only numeric data (so, no strings). :+1:

**Comment by NippunSharma** [comment]
@shrit @rcurtin 
Thanks for taking a look at this. @rcurtin You are correct I will have to use a different overload of `data::Load ` to call boost spirit (which I for some reason overlooked). @shrit I will change the sizes to increase the number of rows. I will share the updated benchmarks along with the code I used. Thanks again.

**Comment by NippunSharma** [comment]
Hi @shrit @rcurtin .
Based on your suggestions and reviews I performed the benchmarks once more. This time the csv files were created with number of rows varying from `1 to 1000000` as suggested by @shrit and the dimensionalities taken are `{5, 15, 25}`. Also as pointed out by @rcurtin I am now using boost spirit within mlpack (by calling the correct overload of `data::Load`). According to the results armadillo's csv parser is much faster than boost spirit. This is the link for the [code](https://github.com/NippunSharma/benchmarks_csv_parsers/blob/main/benchmark.cpp)


**System Configuration:
Operating System: Ubuntu 20.04
Processor: Intel i5 8th gen
RAM: 8GB**

Following are the plots obtained: 

![image](https://user-images.githubusercontent.com/53967069/100092041-7f81cc00-2e7b-11eb-9a86-d78cf1f26953.png)

![image](https://user-images.githubusercontent.com/53967069/100092061-8577ad00-2e7b-11eb-9118-aac4d5c72734.png)

![image](https://user-images.githubusercontent.com/53967069/100092076-88729d80-2e7b-11eb-8098-5ce2f5ee52e6.png)

**Comment by shrit** [comment]
@NippunSharma Thanks for sharing these results, it seems that the spirit parser is close to linear while the armadillo one is `root(n)`
I did not think that the boost::spirit parser is that slow when loading only numbers from a csv file. I have no idea why is that slow, but at least we have an upper threshold when changing boost spirit.

**Comment by rcurtin** [comment]
Wow, I am surprised!  I had thought that our boost::spirit parser was actually pretty fast, but apparently I was mistaken about this.

The big reason that we have the `boost::spirit` parser is that it can handle strings inside of a CSV file, and when it gets a string (or actually I think when it parses anything), it will call out to `DatasetMapper<MappingPolicy>` in order to map that string to a numeric type that we can put into an `arma::mat`.

It seems like the clear choice here is to replace `boost::spirit` based on these numbers, but we would need something that can interact with `DatasetMapper` in the same way.  Reusing the Armadillo code seems like it wouldn't work, because one of the big complexities of parsing strings is nested quotes; so, for instance, this file is valid:

```
1, 2, "3, 4, 5"
3, 4, "hello"
```

but it's hard to parse that right unless you're tracking the quotes.

So at this point, maybe it's worth seeing if a by-hand implementation would be easy, or if we should depend on some other C++ CSV reader?  I linked a few above, but I have no idea how easy they are to work with.

**Comment by NippunSharma** [comment]
Hi @rcurtin @shrit . Sorry that I was not able to devote more time to this as I was caught in other PR's and then my semester exams also came but there is some free time between the exams now and I can devote time to this. From the benchmarks I think it was clear that we should remove the boost spirit dependency. Should I go ahead and read about the csv parsers mentioned by @rcurtin ? Also, I was looking at how boost spirit works and I came to know that it uses Parsing Expression Grammar (if I am not mistaken). So would it be a good idea to replace boost spirit with a library that uses PEG too ? I came across [PEGTL](https://github.com/taocpp/PEGTL) while searching. Are there some more suggestions or libraries ? I think removing boost spirit might reduce the footprint by a large factor.

**Comment by rcurtin** [comment]
No worries @NippunSharma.  I think there's no need for anything as complex as PEGTL.  Really all we are using `boost::spirit` for is a generic CSV parser.  Maybe we can find a simpler solution for that problem, or implement a simple CSV parser by hand. :+1:

**Comment by heisenbuug** [comment]
@NippunSharma hey, do you mind uploading the log file to the repo as well? As you must have noticed on the mailing list I am working on this as my potential GSoC project, it would be a great help, thank you.

**Comment by NippunSharma** [comment]
@heisenbuug Hi, I saw your mail and the benchmark code also. I will have to search through my system for the log files, I will upload them once I find them :+1: . However, I doubt that our results will be the same because of the difference in system configuration. For the plots, on the horizontal axis, I have the number of data points and on the vertical axis, I have the time taken to read in seconds. You can try saving the plot images, then you will be able to see the labels and ticks as well.

**Comment by heisenbuug** [comment]
Yes, actually I was plotting the whole column at once then I realized that I need to separate it into 3 different lists. I think the code explains better. 
Take a look here: https://github.com/heisenbuug/Benchmark-CSV-Parsers/blob/main/csvparser_log_check.ipynb

**Comment by heisenbuug** [comment]
5 columns
![5_columns](https://user-images.githubusercontent.com/32970979/113344763-235e1580-934f-11eb-9321-dd2e05441bf3.png)
15 columns
![15_columns](https://user-images.githubusercontent.com/32970979/113344776-26f19c80-934f-11eb-8871-9ef05f9ee6fc.png)
25 columns
![25_columns](https://user-images.githubusercontent.com/32970979/113344783-28bb6000-934f-11eb-849e-03929cb1297a.png)

**Comment by shrit** [comment]
@heisenbuug Thanks you for sharing these results,  I suppose that the y-axis is equal to the time in seconds? And the x-axis the number of rows?

**Comment by heisenbuug** [comment]
> @heisenbuug Thanks you for sharing these results, I suppose that the y-axis is equal to the time in seconds? And the x-axis the number of rows?

Yup, you are right, sorry I forgot to mention that.

**Comment by shrit** [comment]
@heisenbuug Looking at your code, it seems to me that you are using two for loop to get the values.
However looking at these examples (this one https://github.com/d99kris/rapidcsv/blob/master/examples/ex001.cpp) it might be possible to get the entire column directly.
The return value seems to be `std::vector`, which can be converted to `arma::col`, or `arma::row`
The idea is to loop the columns (which usually is a small number) and recover directly all related rows using one loop.
I guess, this will reduce the parsing time.
I did not read all the API of rapidcsv, but if you can get out a matrix directly from the document that will be great.

**Comment by heisenbuug** [comment]
> @heisenbuug Thanks you for sharing these results, I suppose that the y-axis is equal to the time in seconds? And the x-axis the number of rows?

Yupp, you right...


> @heisenbuug Looking at your code, it seems to me that you are using two for loop to get the values.
> However looking at these examples (this one https://github.com/d99kris/rapidcsv/blob/master/examples/ex001.cpp) it might be possible to get the entire column directly.
> The return value seems to be `std::vector`, which can be converted to `arma::col`, or `arma::row`
> The idea is to loop the columns (which usually is a small number) and recover directly all related rows using one loop.
> I guess, this will reduce the parsing time.
> I did not read all the API of rapidcsv, but if you can get out a matrix directly from the document that will be great.

Yes, I was thinking the same...
I will make the necessary changes and run the benchmark again...

**Comment by shrit** [comment]
Great to hear @heisenbuug, I would also include the other parser (fast-cpp) in the same test, this will save you a lot of time by running only one test.

**Comment by heisenbuug** [comment]
Okay, I will compare all 4 of them and update the graphs.

**Comment by heisenbuug** [comment]
hey @shrit, I noticed something on `fast-csv` repo

> Q: Does the library support string fields that span multiple lines?
> 
> A: No. This feature has been often requested in the past, however, it is difficult to make it work with the current design without breaking something else.

Is it okay for us to not have this feature?

**Comment by shrit** [comment]
@heisenbuug I am not sure if we need it or not, maybe @rcurtin have an idea here.

**Comment by rcurtin** [comment]
I don't think it's important.  I think personally that a CSV file with a line break in the middle of the row is malformed... but the definition of CSV is fuzzy, so it's not really clear whether my opinion is right. :)

@shrit and I were talking today at the video meet-up about this... the Armadillo CSV parser is so fast that if our simulations show that the other alternatives we're comparing aren't as fast as Armadillo, then perhaps we can copy `diskio_meat.hpp` (with attribution of course!) from Armadillo into our code as a CSV parser, and then adapt it accordingly to load categorical data (since Armadillo can't currently do that).

**Comment by heisenbuug** [comment]
Benchmarking for all 4 parsers

- Armadillo's parser
- mlpack's custom CSV parser
- rapidcsv
- fastcsv

Code: https://github.com/heisenbuug/Benchmark-CSV-Parsers

**5 Columns**
![5_cols](https://user-images.githubusercontent.com/32970979/113469647-ef791200-946c-11eb-99e5-2bccc42e3373.png)

**15 Columns**
![15_cols](https://user-images.githubusercontent.com/32970979/113469656-fbfd6a80-946c-11eb-9ff3-8bec516acd17.png)

**25 Columns**
![25_cols](https://user-images.githubusercontent.com/32970979/113469658-03bd0f00-946d-11eb-834e-b649bde327bd.png)

**Comment by rcurtin** [comment]
Nice, thanks for putting this together!  I have some comments about the code:

 * For rapidcsv, we should read in one row at a time.  Reading a column from a CSV is a bit of a strange operation, since on disk a column is not contiguous.  But a row is contiguous.  So, we should read a row at a time, and put that into the matrix.  Also, the matrix should be transposed on load---so the number of columns in the matrix is equal to the number of rows in the CSV, and the number of rows in the matrix is equal to the number of columns in the CSV.

 * With rapidcsv you chose to load an `arma::fmat`, but the other tests use `arma::mat`; we should use the same data type for the comparisons.

 * I don't follow the FastCSV implementation.  It looks like you are just reading a row at a time, then extracting from a stringstream?  That could easily be done with an fstream; no FastCSV needed.  Surely the library gives better support for parsing elements?

I'm not sure whether any of those changes will change the actual results you're seeing here, but I think it would be worthwhile to go through and fix these issues so that we know our comparison is good.  After all, if we're going to make a big decision on what CSV parser to use, we need to make that decision on the best data we can. :)

**Comment by heisenbuug** [comment]
> * I don't follow the FastCSV implementation.  It looks like you are just reading a row at a time, then extracting from a stringstream?  That could easily be done with an fstream; no FastCSV needed.  

For `FastCSV` the `.next_line()` function returns a char pointer which also includes comma as an element so I separated them based on comma and inserted each one into a vector(after converting them to double). 

>Surely the library gives better support for parsing elements?

FastCSV doesn't have well-documented examples so I read the code and figured this out. I myself thought the same that I can use a simple `fstream` to read data like that but I couldn't figure out any other way. I even found it strange to have `number of columns` as a template parameter. 

I also opened an issue on that repo but no response yet. How shall I work around it then?

I will make the other necessary changes accordingly and run the benchmark again.

Let me know what we can do about `FastCSV` coz if we plan on not using it I will not include it in the benchmark. Is that okay?

**Comment by heisenbuug** [comment]
Heyy @rcurtin I made the changes

- Now I am loading all the data in fmat
- transposing the matrix after loading
- Still in doubt regarding the `FastCSV` but I am still including it in the benchmarks

I will post the results once its done.

**Comment by rcurtin** [comment]
If FastCSV is too difficult to figure out how to use correctly, well, that's kind of something that might rule it out. :)

**Comment by heisenbuug** [comment]
Hey @rcurtin, these are new graphs
![5_cols](https://user-images.githubusercontent.com/32970979/113502458-0ccfde00-954a-11eb-9850-592547c36f1b.png)
![15_cols](https://user-images.githubusercontent.com/32970979/113502459-0e010b00-954a-11eb-9548-97017d0a8974.png)
![25_cols](https://user-images.githubusercontent.com/32970979/113502460-0e99a180-954a-11eb-91e4-294c898f6ba8.png)

> Also, I have another doubt, if we decide to go with copying armadillo's parser into mlpack, then we have to work on categorical data. Currently, we are doing that using DatasetMapper if I am not wrong, and DatasetMapper was working based on the custom CSV parser's implementation in load_csv files. So we basically have to extend the arma'a parser in such a way that it can support the existing implementation of DatasetMapper.

Can you comment on this, thank you.

**Comment by shrit** [comment]
@heisenbuug @rcurtin, it is the same performance as before, I would vote for armadillo parser. You would have to copy the `diskio_meat.hpp` from armadillo and adapt it for mlpack. I do think that it is impossible to get better performance than `armadillo`.
These are my thoughts, let me know what do you think.

**Comment by rcurtin** [comment]
Agreed, @shrit's idea seems like the right call here.  Basically we can adapt the CSV reader in `diskio_meat.hpp` to also have functionality to map non-numeric types.  Here's the place where Armadillo's CSV reader converts a string into a numeric type: https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.4.x/include/armadillo_bits/diskio_meat.hpp#L1680.  In essence, we would want to call `DatasetMapper::MapString()` at this location.  That is just one idea for how to do it, but in the end the solution for non-numeric data will probably look something like that.

@conradsnicta just an FYI---you might be pleased with the results above; the Armadillo CSV parser is blazing fast.

**Comment by heisenbuug** [comment]
Thank you for the feedback and inputs @shrit and @rcurtin 
Now that we have decided that we will be going with `diskio_meat.hpp` I will make my proposal around adapting this for non-numeric data. We basically have to use the functionality of  `DatasetMapper` to get `diskio_meat.hpp` up and working for non-numeric data.

> Here's the place where Armadillo's CSV reader converts a string into a numeric type: https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.4.x/include/armadillo_bits/diskio_meat.hpp#L1680. In essence, we would want to call DatasetMapper::MapString() at this location.

@rcurtin I will make necessary changes here so that we can see how it works, and we can also have a small working prototype-like code to see how this may turn out in the future.

**Comment by conradsnicta** [comment]
> @conradsnicta just an FYI---you might be pleased with the results above; the Armadillo CSV parser is blazing fast.

@rcurtin That's actually worringly surprising :) I rewrote it a while back, and my recollection is that there is still quite a lot of room for improvement.  I'm not entirely proud of the resulting code.  I do hope that fast != broken, because the speed difference compared to the other CSV readers is rather suspicious.

Maybe it's simply due to supporting only numerical data, but my gut tells me something smells fishy here.  Can we double check that the data read by the Armadillo CSV reader actually matches the data read by the other CSV readers?

While I took care for correctness and did many rounds of testing, having more eyes on the code would be useful.  As the saying goes: when something looks too good to be true ...

**Comment by heisenbuug** [comment]
@rcurtin @shrit I have a doubt.
I believe that the proper way to include `diskio_meat.hpp` would be:

- mlpack's program will use `diskio_meat_mlpack.hpp` and this file should be aware of armadillo files. 
- Also about the armadillo's inbuild parser i.e. `diskio_meat.hpp`, I think it can stay there, it should not make any difference, right?

I am stuck a bit in first point, what I did is:
- placed `diskio_meat_mlpack` in my working directory.
- went to my armadillo installation and changed this [line](https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.4.x/include/armadillo#L635) to include `diskio_meat_mlpack.hpp

But I am getting some error, it should be regarding the linking of files. I think `diskio_meat_mlpack.hpp` is not able to access armadillo files, I included the path of  `armadillo_bits` in `task.json` file of vscode, but it's still giving an error.

These are last few lines of the error terminal output
```
/home/heisenbug/Desktop/test/diskio_meat_mlpack.hpp:5066:36: error: expected ‘,’ or ‘...’ before ‘<’ token
 5066 | diskio::save_ppm_binary(const field<T1>& x, std::ostream& f)
      |                                    ^
/home/heisenbug/Desktop/test/diskio_meat_mlpack.hpp:5066:1: error: redefinition of ‘template<class T1> bool save_ppm_binary(int)’
 5066 | diskio::save_ppm_binary(const field<T1>& x, std::ostream& f)
      | ^~~~~~
/home/heisenbug/Desktop/test/diskio_meat_mlpack.hpp:4839:1: note: ‘template<class eT> bool save_ppm_binary(int)’ previously declared here
 4839 | diskio::save_ppm_binary(const Cube<eT>& x, const std::string& final_name)
      | ^~~~~~
```
What should I do now? Where am I going wrong?

**Comment by shrit** [comment]
@heisenbuug 
1. I would first copy the file from armadillo to mlpack
2. Add at the top of the file that it is originated from armadillo and adapted for mlpack.
3. Include these function in `mlpack::data` name space or similar

**Comment by rcurtin** [comment]
Don't forget that if you add a new file to mlpack, the `CMakeLists.txt` in that directory needs to be updated. :+1:

Inside the file, I would only keep the functions that are necessary for the task at hand.  The Armadillo sources are inside the `arma` namespace, so you may need to add `arma::` throughout to the different types.  (For instance, `field<T1>` should become `arma::field<T1>`.)

**Comment by heisenbuug** [comment]
Sorry for the late update, I am a bit busy with my exams actually.

@shrit I was able to successfully follow up on what you guys advised and included the parser in `mlpack::data` namespace.

@rcurtin as you suggested I only copied the required fucntions and after that adding `arma::` did the job.
For now, I only copied few functions to load data and also understood how `convert_token` is working. I will replace it with our `mlpack::data::DatasetMapper()`. I will update once it's done. 

Shall I open a PR for this, or should I wait still I integrate the `mlpack::data::DatasetMapper()`

**Comment by shrit** [comment]
@heisenbuug I think you can open a PR so we can continue the discussion there.
